cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,544 @@
+ * for that key is created and that is put in the store too. The system then flushes
+ *  the keystore to the file system.
+ *  During the flush, the current file is first backed up (_OLD). Then the keystore
+ *  is written to temporary file (_NEW). If that is successful then the temporary
+ *  file is renamed to the secure store file. If anything fails during this process
+ *  then the keystore reverts to the last  successfully written file.
+ *
+ *  The keystore is flushed to the filesystem after every put and delete.
+ */
+@Singleton
+class FileSecureStore implements SecureStore, SecureStoreManager {
+  private static final Logger LOG = LoggerFactory.getLogger(FileSecureStore.class);
+
+  private static final String SCHEME_NAME = "jceks";
+  /*
+   This is used to create the key for storing metadata in the keystore. Changing this will make the
+   metadata for existing entries unreachable.
+   */
+  private static final String METADATA_SUFFIX = "_metadata";
+  private static final String DEFAULT_FILE_PATH = "/tmp";
+  private static final String DEFAULT_FILE_NAME = "securestore";
+  private static final char[] DEFAULT_PASSWORD = "cdapsecret".toCharArray();
+  /*
+   Java Keystore needs an algorithm name to store a key, it is not used for any checks, only stored,
+   since we are not handling the encryption we don't care about this.
+  */
+  private static final String ALGORITHM_PROXY = "none";
+
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+
+  private KeyStore keyStore;
+  private boolean changed = false;
+
+  FileSecureStore(CConfiguration cConf) throws IOException {
+    // Get the path to the keystore file
+    String pathString = cConf.get(Constants.Security.Store.FILE_PATH, DEFAULT_FILE_PATH);
+    Path dir = Paths.get(pathString);
+    path = dir.resolve(DEFAULT_FILE_NAME);
+
+    // Get the keystore password
+    String passwordString = cConf.get(Constants.Security.Store.FILE_PASSWORD);
+    if (passwordString == null || passwordString.isEmpty()) {
+      password = DEFAULT_PASSWORD;
+    } else {
+      password = passwordString.toCharArray();
+    }
+
+    locateKeystore();
+    ReadWriteLock lock = new ReentrantReadWriteLock(true);
+    readLock = lock.readLock();
+    writeLock = lock.writeLock();
+  }
+
+  /**
+   * Stores an element in the secure store. If the key already exists then delete it first.
+   * @param name Name of the element to store
+   * @param data The data that needs to be securely stored
+   * @param properties Metadata associated with the data
+   * @throws IOException
+   */
+  @Override
+  public void put(String name, byte[] data, Map<String, String> properties) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    writeLock.lock();
+    try {
+      if (keyStore.containsAlias(name)) {
+        // Clear the existing key so that we can write the new one.
+        delete(name);
```

Comment:
Is it necessary? Also, this mean if the put failed, then the old entry is also gone?